dll/win32/samsrv/samrpc.c

```diff
@@ -397,16 +397,16 @@ SamrQuerySecurityObject(IN SAMPR_HANDLE ObjectHandle,
 
     /* Invalidate the SD information that was not requested */
     if (!(SecurityInformation & OWNER_SECURITY_INFORMATION))
-        ((PISECURITY_DESCRIPTOR)RelativeSd)->Owner = NULL;
+        RelativeSd->Owner = 0;
 
     if (!(SecurityInformation & GROUP_SECURITY_INFORMATION))
-        ((PISECURITY_DESCRIPTOR)RelativeSd)->Group = NULL;
+        RelativeSd->Group = 0;
 
     if (!(SecurityInformation & DACL_SECURITY_INFORMATION))
-        ((PISECURITY_DESCRIPTOR)RelativeSd)->Control &= ~SE_DACL_PRESENT;
+        RelativeSd->Control &= ~SE_DACL_PRESENT;
 
     if (!(SecurityInformation & SACL_SECURITY_INFORMATION))
-        ((PISECURITY_DESCRIPTOR)RelativeSd)->Control &= ~SE_SACL_PRESENT;
+        RelativeSd->Control &= ~SE_SACL_PRESENT;
 
     /* Calculate the required SD size */
     Status = RtlMakeSelfRelativeSD(RelativeSd,
```

Comment:
(This comments also applies to the other RtlMakeSelfRelativeSD call.)
Note that usually this function takes as the first parameter a (potentially absolute) SD, and returns via its second parameter a relative SD.
In its internal implementation, https://git.reactos.org/?p=reactos.git;a=blob;f=sdk/lib/rtl/sd.c;hb=88808bad8a29213bc51ca54db265e9425a457981#l645 ,
the first and second parameters are respectively cast as follows:
```c
 655     PISECURITY_DESCRIPTOR Sd = (PISECURITY_DESCRIPTOR)AbsoluteSD; // The first parameter
 656     PISECURITY_DESCRIPTOR_RELATIVE RelSd = (PISECURITY_DESCRIPTOR_RELATIVE)SelfRelativeSD; // The second parameter.
```
So my question is, whether these fixes are correct. Indeed, we have these definitions for the SECURITY_DESCRIPTOR and SECURITY_DESCRIPTOR_RELATIVE:
https://git.reactos.org/?p=reactos.git;a=blob;f=sdk/include/xdk/setypes.h;hb=88808bad8a29213bc51ca54db265e9425a457981#l836
and I wonder whether how the Owner, Group, (and Sacl, Dacl) fields behave in 64-bits compilation.
Indeed, in SECURITY_DESCRIPTORS, these are pointers, while in the _RELATIVE version, these are mere 32-bit ULONGs.

Reply:
The only reason for the cast is that `RtlpQuerySecurityDescriptor` takes `PISECURITY_DESCRIPTOR` as parameter rather than `PSECURITY_DESCRIPTOR`. It should take the latter though, because it handles both absolute and relative.

- [ ] Make RtlpQuerySecurityDescriptor, SepGetOwnerFromDescriptor, etc. take a PSECURITY_DESCRIPTOR as argument.

Reply:
> Rename parameter to RtlMakeSelfRelativeSD from AbsoluteSD to SecurityDescriptor

Hmmm... This would imply that passing a relative security descriptor as the 1st parameter to that function would work, though (on Windows). Is it really the case?

Reply:
Yes

Reply:
Passing a relative descriptor (denoted by `SE_SELF_RELATIVE`) is correct. By nature a relative descriptor contains the security information whereas an absolute one doesn't. With the latter, it actually points to the information instead.

At best `RtlMakeSelfRelativeSD` will just simply duplicate a relative descriptor for you from an already existing relative descriptor. This dance cannot be done the same for `RtlSelfRelativeToAbsoluteSD` though because otherwise you would end up with an absolute security descriptor with pointers to the information of another absolute descriptor with pointers to the information. And we have a rule for that.
https://git.reactos.org/?p=reactos.git;a=blob;f=sdk/lib/rtl/sd.c;hb=aa84645bccda8d7095f6053cae67b1e9bf816345#l762

So I agree with Timo's changes here.